In [1]:
import math
import numpy as np
from scipy import stats
from scipy.stats import t, f, nct, ttest_rel, wilcoxon
from scipy.integrate import quad

<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q1. One-Sample t-Test (Mean)**

A sample of 20 students scored: mean = 72, s = 8. The national average is 70. At α = 0.05, is this school's performance significantly different? Use a one-sample t-test. Compute t, df, the critical value, and state your conclusion.

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ = 70 (school mean equals the national average)  
H1: μ ≠ 70 (school mean is different), two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| t-statistic | $t = \dfrac{\bar{x} - \mu_0}{s / \sqrt{n}}$ |
| df | $n - 1$ |
| 95% CI | $\bar{x} \pm t_{\alpha/2,\,df} \cdot SE$ |

Population SD is unknown and n = 20 is small, so we use the t-distribution with heavier tails.

</div>


In [2]:
n = 20          # sample size
xbar = 72       # sample mean
s = 8           # sample standard deviation
mu = 70         # null hypothesis population mean
alpha = 0.05    # significance level
df = n - 1     # degrees of freedom

In [3]:
t_stat = (xbar - mu) / (s / np.sqrt(n))  # one-sample t-statistic
t_stat

np.float64(1.118033988749895)

In [4]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=19
t_crit

np.float64(2.0930240544083087)

**|t| = 1.118 < t_crit = 2.093, Fail to reject H0.**


In [5]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.27749013957576074)

**p-value = 0.2775 > 0.05, Fail to reject H0.**


In [6]:
se = s / np.sqrt(n)       # standard error
lo = xbar - t_crit * se   # lower bound of 95% CI
hi = xbar + t_crit * se   # upper bound of 95% CI
lo, hi

(np.float64(68.25588474864064), np.float64(75.74411525135936))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Conclusion**

t_stat (1.118) < t_crit (2.093) &nbsp;&nbsp;|&nbsp;&nbsp; p-value (0.2775) > alpha (0.05)

**Fail to reject H0.** The sample mean of 72 is not significantly different from the national average of 70; the 95% CI (68.26, 75.74) contains 70.

</div>


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q2. Two-Sample t-Test (Pooled, Equal Variances)**

Drug Group (n = 15, x̄ = 5.2, s = 1.1) vs Placebo Group (n = 15, x̄ = 4.8, s = 0.9). Test whether the drug has a significant effect at α = 0.05. Assume equal variances. Compute the pooled standard deviation and the t-statistic manually.

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ1 = μ2 (drug and placebo have the same mean)  
H1: μ1 ≠ μ2 (drug group mean differs), two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| Pooled SD | $s_p = \sqrt{\dfrac{(n_1-1)s_1^2+(n_2-1)s_2^2}{n_1+n_2-2}}$ |
| t-statistic | $t = \dfrac{\bar{x}_1 - \bar{x}_2}{s_p\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}}$ |
| df | $n_1 + n_2 - 2 = 28$ |

Equal variances are assumed, so both samples share a common variance estimate, pooling gives a tighter SE than keeping them separate.

</div>


In [7]:
na = 15          # drug group sample size
xbara = 5.2      # drug group sample mean
sa = 1.1         # drug group sample standard deviation
nb = 15          # placebo group sample size
xbarb = 4.8      # placebo group sample mean
sb = 0.9         # placebo group sample standard deviation
alpha = 0.05     # significance level
df = na + nb - 2  # degrees of freedom for pooled two-sample t-test

In [8]:
sp = math.sqrt(((na - 1) * sa**2 + (nb - 1) * sb**2) / (na + nb - 2))  # pooled standard deviation
sp

1.004987562112089

In [9]:
se = sp * np.sqrt(1 / na + 1 / nb)   # pooled standard error
t_stat = (xbara - xbarb) / se              # two-sample pooled t-statistic
t_stat

np.float64(1.0900086292691404)

In [10]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=28
t_crit

np.float64(2.048407141795245)

**|t| = 1.090 < t_crit = 2.048, Fail to reject H0.**


In [11]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.2850015227710626)

**p-value = 0.2850 > 0.05, Fail to reject H0.**


In [12]:
diff = xbara - xbarb         # observed difference in means
lo = diff - t_crit * se      # lower bound of 95% CI
hi = diff + t_crit * se      # upper bound of 95% CI
lo, hi

(np.float64(-0.35170309180716086), np.float64(1.1517030918071616))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Conclusion**

t_stat (1.090) < t_crit (2.048) &nbsp;&nbsp;|&nbsp;&nbsp; p-value (0.2850) > alpha (0.05)

**Fail to reject H0.** No significant difference between drug and placebo groups; the 95% CI (-0.35, 1.15) spans zero.

</div>


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q3. Welch's t-Test vs Student's t-Test**

Group A: n1 = 12, x̄1 = 45.3, s1 = 8.7 &nbsp;|&nbsp; Group B: n2 = 30, x̄2 = 41.8, s2 = 15.2  
(a) Check equality of variances (variance ratio) and choose the appropriate test with justification.  
(b) Compute Welch's df and perform a two-sided Welch's t-test at α = 0.05.  
(c) Construct a 95% CI for (μ1 - μ2).  
(d) State the modern consensus on when to use Welch's t-test.

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ1 = μ2 (Group A and B have the same mean)  
H1: μ1 ≠ μ2, two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| Variance ratio | $\max(s_1^2,\,s_2^2) / \min(s_1^2,\,s_2^2)$ |
| Welch SE | $SE = \sqrt{s_1^2/n_1 + s_2^2/n_2}$ |
| Welch df | $df = (r_1+r_2)^2 / (r_1^2/(n_1-1) + r_2^2/(n_2-1))$ |

Variance ratio > 2 means equal-variance assumption is untenable; Welch's test adjusts df via Welch-Satterthwaite so it does not pool.

</div>


In [13]:
na = 12        # Group A sample size
xa = 45.3      # Group A sample mean
sa = 8.7       # Group A sample standard deviation
nb = 30        # Group B sample size
xb = 41.8      # Group B sample mean
sb = 15.2      # Group B sample standard deviation
alpha = 0.05   # significance level

#### Part (a) : Variance Check


In [14]:
vara = sa**2                              # Group A variance
varb = sb**2                              # Group B variance
ratio = max(vara, varb) / min(vara, varb)  # variance ratio
ratio

3.0524507861012027

In [15]:
# Pick test
test_choice = "Welch's t-test" if ratio > 2 else "Student's t-test"
ratio, test_choice

(3.0524507861012027, "Welch's t-test")

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (a) Conclusion**

Variance ratio = 3.05 > 2, variances are unequal.

**Use Welch's t-test.** The equal-variance assumption is not tenable; pooling would distort the SE.

</div>


#### Part (b) : Welch's t-Test


In [16]:
se = np.sqrt(sa**2 / na + sb**2 / nb)  # Welch standard error
t_stat = (xa - xb) / se                        # Welch t-statistic
t_stat

np.float64(0.9351193853114768)

In [17]:
num = (vara / na + varb / nb)**2                          # numerator of Welch-Satterthwaite formula
den = (vara / na)**2 / (na - 1) + (varb / nb)**2 / (nb - 1)  # denominator
df = num / den                                               # raw degrees of freedom
df_ceil = math.ceil(df)                                       # round up per convention
df_ceil

35

In [18]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05
t_crit

np.float64(2.0308187174932497)

**|t| = 0.935 < t_crit = 2.031, Fail to reject H0.**


In [19]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.3561964916843626)

**p-value = 0.3562 > 0.05, Fail to reject H0.**


<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (b) Conclusion**

t_stat (0.935) < t_crit (2.031) &nbsp;&nbsp;|&nbsp;&nbsp; p-value (0.3562) > alpha (0.05)

**Fail to reject H0.** No significant difference between group means at α = 0.05.

</div>


#### Part (c) : 95% Confidence Interval


In [20]:
margin = t_crit * se           # margin of error
lo = (xa - xb) - margin        # lower bound of 95% CI
hi = (xa - xb) + margin        # upper bound of 95% CI
lo, hi

(np.float64(-4.101024663667764), np.float64(11.101024663667765))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (c) Conclusion**

95% CI for (μ1 - μ2): **(−4.10, 11.10)**, contains zero, consistent with failing to reject H0.

</div>


#### Part (d) : Modern Consensus on Welch's t-Test


The modern consensus, as summarized by Delacre, Lakens, and Leys (2017), is to default to Welch's t-test rather than testing for equal variances first and then choosing. The reasons are:

- When variances are equal, Welch's t-test performs nearly as well as Student's t-test with negligible loss of power.
- When variances are unequal, Student's t-test can produce inflated Type I error rates, especially when sample sizes also differ.
- Pre-testing for equal variances (such as Levene's or Bartlett's) adds an extra layer of uncertainty and does not reliably protect against the error inflation in Student's test.
- Welch's t-test is robust across both equal and unequal variance conditions, making it the safer all-purpose choice.

The practical takeaway: unless there is a domain-specific reason to assume equal variances, Welch's t-test is the appropriate default for comparing two independent group means.


<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (d) Conclusion**

Welch's t-test is the safer default, it costs almost nothing when variances are equal and protects Type I error control when they are not.

</div>


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q4. Paired t-Test vs Independent t-Test**

25 students were tested **before** and **after** a coaching program: Before x̄ = 68, s = 10; After x̄ = 74, s = 12; r = 0.72, n = 25.  
(a) Compute SE of the mean difference using the paired formula.  
(b) Run a paired t-test, then repeat as an independent t-test. Compare both.  
(c) Show quantitatively how much power is lost by ignoring the correlation.

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ_d = 0 (coaching has no effect)  
H1: μ_d ≠ 0, two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| Paired SE | $SE_d = \sqrt{s_1^2 + s_2^2 - 2r s_1 s_2}\,/\,\sqrt{n}$ |
| Paired t | $t = \bar{d}\,/\,SE_d$, df = n-1 |
| Independent SE | $SE_{\text{ind}} = \sqrt{s_1^2/n + s_2^2/n}$ |

Paired design removes between-student variability; the correlation term $2r s_1 s_2$ reduces SE, increasing power, ignoring it inflates SE by the SE ratio.

</div>


In [21]:
n    = 25    # number of pairs (students)
xba  = 68    # before-group mean
sa   = 10    # before-group standard deviation
xbb  = 74    # after-group mean
sb   = 12    # after-group standard deviation
r    = 0.72  # Pearson correlation between paired observations
alpha = 0.05 # significance level
d    = xbb - xba  # mean difference
df   = n - 1       # degrees of freedom for paired test
d

6

#### Part (a) : SE of the Mean Difference


In [22]:
sep = np.sqrt(sa**2 + sb**2 - 2 * r * sa * sb) / np.sqrt(n)  # paired SE
sep

np.float64(1.687601848778319)

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (a) Conclusion**

Paired SE = 1.687. The correlation of 0.72 removes 172.8 out of 244 units of combined variance, roughly halving the SE compared to an independent-samples formula.

</div>


#### Part (b) : Paired t-Test


In [23]:
tp = d / sep  # paired t-statistic: mean difference divided by paired SE
tp

np.float64(3.555340973549829)

In [24]:
t_crit_p = t.ppf(1 - alpha / 2, df)  # two-tailed critical value, df=24
t_crit_p

np.float64(2.0638985616280245)

**|t| = 3.555 > t_crit = 2.064, Reject H0.**


In [25]:
pvalp = 2 * t.sf(abs(tp), df)  # two-tailed p-value, df=24
pvalp

np.float64(0.0016063038416739707)

**p-value = 0.0016 < 0.05, Reject H0.**


In [26]:
lop = d - t_crit_p * sep  # lower bound of 95% CI for mean difference
hip = d + t_crit_p * sep  # upper bound of 95% CI for mean difference
lop, hip

(np.float64(2.516960971705632), np.float64(9.483039028294368))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Paired t-Test Conclusion**

t_stat (3.555) > t_crit (2.064) &nbsp;&nbsp;|&nbsp;&nbsp; p-value (0.0016) < alpha (0.05)

**Reject H0.** The paired test detects a significant improvement; 95% CI (2.52, 9.48) lies entirely above zero.

</div>


#### Part (b) continued : Independent t-Test (Ignoring Pairing)


In [27]:
sei  = np.sqrt(sa**2 / n + sb**2 / n)  # independent SE, no covariance term
dfi  = n + n - 2                           # degrees of freedom for two-sample test
ti   = d / sei                             # independent t-statistic
sei, ti

(np.float64(3.1240998703626617), np.float64(1.9205531989934397))

In [28]:
t_crit_i = t.ppf(1 - alpha / 2, dfi)  # two-tailed critical value, df=48
t_crit_i

np.float64(2.010634757624232)

**|t| = 1.921 < t_crit = 2.011, Fail to reject H0.**


In [29]:
pvali = 2 * t.sf(abs(ti), dfi)  # two-tailed p-value, df=48
pvali

np.float64(0.06073954082573264)

**p-value = 0.0607 > 0.05, Fail to reject H0.**


In [30]:
loi = d - t_crit_i * sei  # lower bound of 95% CI (independent)
hii = d + t_crit_i * sei  # upper bound of 95% CI (independent)
loi, hii

(np.float64(-0.2814237856405244), np.float64(12.281423785640524))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Full Comparison**

| | Paired t-test | Independent t-test |
|---|---|---|
| SE | 1.687 | 3.124 |
| t-statistic | 3.555 | 1.921 |
| p-value | 0.0016 | 0.0607 |
| Decision | **Reject H0** | **Fail to reject H0** |

Same data, same effect, opposite conclusions, the independent test discards the correlation and inflates SE by 1.85x.

</div>


#### Part (c) : Power Loss from Ignoring Correlation


In [31]:
ratio_se = sei / sep   # SE inflation factor when pairing is ignored
ratio_t  = tp  / ti   # t-statistic deflation factor when pairing is ignored
ratio_se, ratio_t

(np.float64(1.8512067124270133), np.float64(1.8512067124270133))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (c) Conclusion**

Ignoring the correlation inflates SE by 1.85x and deflates the t-statistic by the same factor, turning p < 0.001 into a borderline p = 0.063.

</div>


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q5. Bootstrap vs t-Test**

Hospital wait times (n = 12): [23, 45, 31, 28, 67, 19, 41, 55, 38, 29, 47, 33]. Claimed average = 35 min.  
(a) Run a one-sample t-test.  
(b) Bootstrap hypothesis test: shift data to mean = 35, resample 10,000 times.  
(c) Compare the two p-values. When and why do they diverge?  
(d) Identify the potential outlier and discuss its influence.

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ = 35 (true mean wait time equals the claimed average)  
H1: μ ≠ 35, two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| t-statistic | $t = (\bar{x} - \mu_0)\,/\,(s/\sqrt{n})$, df = n-1 |
| Bootstrap shift | $l_{\text{shift}} = l - \bar{x} + \mu_0$ |
| Bootstrap p | proportion of $|\bar{x}^*_b - \mu_0| \geq |\bar{x} - \mu_0|$ |

One sample, unknown σ, n = 12, t-distribution is appropriate. Bootstrap provides a distribution-free cross-check since normality is unverifiable at this sample size.

</div>


#### Part (a) : One-Sample t-Test


In [32]:
l = np.array([23, 45, 31, 28, 67, 19, 41, 55, 38, 29, 47, 33])  # wait time data
n = 12            # sample size
mu = 35           # claimed population mean
df = n - 1        # degrees of freedom
alpha = 0.05      # significance level
x = np.mean(l)    # sample mean
x

np.float64(38.0)

In [33]:
std = np.std(l, ddof=1)   # sample standard deviation, ddof=1 for unbiased estimate
se = std / np.sqrt(n)     # standard error of the mean
se

np.float64(3.998105612017832)

In [34]:
t_stat = (x - mu) / se  # one-sample t-statistic
t_stat

np.float64(0.7503553660469486)

In [35]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=11
t_crit

np.float64(2.200985160091639)

**|t| = 0.750 < t_crit = 2.201, Fail to reject H0.**


In [36]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.46878728507546324)

**p-value = 0.4688 > 0.05, Fail to reject H0.**


In [37]:
moe   = t_crit * se  # margin of error
lower = x - moe      # lower bound of 95% CI
upper = x + moe      # upper bound of 95% CI
lower, upper

(np.float64(29.200228879469652), np.float64(46.79977112053035))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (a) Conclusion**

t_stat (0.750) < t_crit (2.201) &nbsp;&nbsp;|&nbsp;&nbsp; p-value (0.4688) > alpha (0.05)

**Fail to reject H0.** Sample mean of 38 is not significantly different from the claimed 35; 95% CI (29.20, 46.80) contains 35.

</div>


#### Part (b) : Bootstrap Hypothesis Test


In [38]:
np.random.seed(0)         # fix seed for reproducibility
shifted = l - x + mu      # shift data so mean equals mu, preserving spread
np.mean(shifted)           # verify shifted mean equals 35

np.float64(35.0)

In [39]:
n_boot = 10000                                              # number of bootstrap resamples
boot_samples = np.random.choice(shifted, size=(n_boot, n), replace=True)  # resample with replacement
boot_means = boot_samples.mean(axis=1)                      # mean of each resample
boot_means.shape

(10000,)

In [40]:
obs_diff = abs(x - mu)                              # observed deviation from null mean
extreme = np.abs(boot_means - mu) >= obs_diff       # bootstrap means at least as extreme
p_val_boot = np.mean(extreme)                        # proportion of extreme samples
p_val_boot

np.float64(0.44)

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (b) Conclusion**

Bootstrap p-value ≈ 0.44, Fail to reject H0. About 44% of 10,000 resamples produced a deviation as large as the observed 3.0, confirming the mean of 38 is not unusual under H0.

</div>


#### Part (c) : Comparison


In [41]:
p_val, p_val_boot  # t-test p-value vs bootstrap p-value

(np.float64(0.46878728507546324), np.float64(0.44))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (c) Conclusion**

| Method | p-value | Decision |
|---|---|---|
| One-sample t-test | 0.4688 | Fail to reject H0 |
| Bootstrap test | ~0.44 | Fail to reject H0 |

Both agree here. They diverge most when normality breaks down (heavy tails, skew, small n with outliers), in those cases the bootstrap result is more trustworthy.

</div>


#### Part (d) : Outlier Detection and Influence


In [42]:
q1 = np.percentile(l, 25)      # first quartile
q3 = np.percentile(l, 75)      # third quartile
iqr = q3 - q1                  # interquartile range
lower_bound = q1 - 1.5 * iqr   # lower IQR fence
upper_bound = q3 + 1.5 * iqr   # upper IQR fence
lower_bound, upper_bound

(np.float64(3.625), np.float64(70.625))

In [43]:
outliers        = l[(l < lower_bound) | (l > upper_bound)]  # values outside IQR fences
mean_without_67 = np.mean(l[l != 67])                       # mean after removing extreme value
outliers, l.max(), mean_without_67

(array([], dtype=int64), np.int64(67), np.float64(35.36363636363637))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (d) Conclusion**

No formal IQR outliers (upper fence = 70.63). However, value 67 pulls the mean from 35.36 (without it) to 38.0, a shift of 2.6 min. In a small sample of 12, a single high observation carries meaningful leverage.

</div>


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q6. Effect Size vs Statistical Significance**

Two t-tests both yield **p = 0.03**.  
| | Test 1 | Test 2 |
|---|---|---|
| n per group | 8 | 800 |
| Mean difference | 5 | 5 |
| Pooled SD | 20 | 20 |

(a) Compute Cohen's d for each.  (b) Compute statistical power for each at α = 0.05.  (c) What does the p-value conflate that Cohen's d separates?

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ1 = μ2 &nbsp;|&nbsp; H1: μ1 ≠ μ2, two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| Cohen's d | $d = (\bar{x}_1 - \bar{x}_2)\,/\,s_p$ |
| NCP | $\delta = d \cdot \sqrt{n/2}$ |
| Power | $P(T_{df,\,\delta} > t_{\text{crit}}) + P(T_{df,\,\delta} < -t_{\text{crit}})$ |

Cohen's d is pure effect size (no n); the t-statistic = d * sqrt(n/2), so the p-value mixes effect and sample size into one number.

</div>


In [44]:
mean_diff = 5   # observed mean difference (same for both tests)
sp        = 20  # pooled standard deviation (same for both tests)
alpha     = 0.05

#### Part (a) : Cohen's d


In [45]:
d = mean_diff / sp  # Cohen's d — same for both tests because mean_diff and sp are identical
d

0.25

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (a) Conclusion**

Cohen's d = 0.25 for both tests, identical, because effect size does not depend on n.

</div>


#### Part (b) : Statistical Power


In [46]:
# Test 1: n = 8 per group
n1      = 8
df1     = 2 * n1 - 2                      # degrees of freedom
se1     = sp * np.sqrt(2 / n1)            # SE of mean difference
t_stat1 = mean_diff / se1                 # t-statistic under H1
t_crit1 = t.ppf(1 - alpha / 2, df1)      # two-tailed critical value
ncp1    = t_stat1                         # non-centrality parameter
power1  = nct.sf(t_crit1, df1, ncp1) + nct.cdf(-t_crit1, df1, ncp1)  # two-tailed power
power1

np.float64(0.07529185896013754)

In [47]:
# Test 2: n = 800 per group
n2      = 800
df2     = 2 * n2 - 2
se2     = sp * np.sqrt(2 / n2)
t_stat2 = mean_diff / se2
t_crit2 = t.ppf(1 - alpha / 2, df2)
ncp2    = t_stat2
power2  = nct.sf(t_crit2, df2, ncp2) + nct.cdf(-t_crit2, df2, ncp2)
power2

np.float64(0.9988053879982561)

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (b) Conclusion**

| | Test 1 (n=8) | Test 2 (n=800) |
|---|---|---|
| Power | 0.0753 | 0.9988 |

Same d = 0.25, yet power jumps from 7.5% to 99.9%, entirely driven by n through the SE.

</div>


#### Part (c) : What the p-value Conflates That Cohen's d Separates

**The key insight**

The t-statistic is literally a product of two things:

$$t = \underbrace{\frac{\bar{x}_1 - \bar{x}_2}{s_p}}_{\text{effect size (Cohen's d)}} \times \underbrace{\sqrt{\frac{n}{2}}}_{\text{sample size term}}$$

The p-value just converts this single number to a probability. It can't separate which factor pushed t up. Cohen's d keeps only the first factor, so it purely measures the signal-to-noise ratio in the population, with no n involved.

**The p-value conflates both. Cohen's d separates them.**

| Scenario | p-value says | Cohen's d says | Reality |
|---|---|---|---|
| Large n, tiny effect | Significant (p < 0.05) | Small (d ≈ 0.01) | Statistically significant, practically irrelevant |
| Small n, large effect | Not significant (p > 0.05) | Large (d ≈ 0.8) | Practically important, underpowered study |
| Large n, large effect | Highly significant | Large | Both agree |

This is why modern statistics requires reporting effect sizes alongside p-values: they answer different questions.


<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (c) Conclusion**

The p-value conflates **effect magnitude** (Cohen's d) and **sample size** into one number. Cohen's d strips out n entirely, a large n can make a trivially small d = 0.01 produce p < 0.05. Always report effect size alongside p-values.

</div>


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q7. Permutation Test**

Salaries ($k): Engineering [95, 102, 88, 110, 98, 105, 92] vs Marketing [78, 82, 75, 91, 85, 79].  
(a) Run a standard two-sample Welch's t-test.  
(b) Run a permutation test with 5,000 permutations.  
(c) With n = 7 and n = 6, which test is more valid and why?  
(d) What does 'assumption-free' actually mean for the permutation test?

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ_eng = μ_mkt &nbsp;|&nbsp; H1: μ_eng ≠ μ_mkt, two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| Welch t | $t = (\bar{x}_1 - \bar{x}_2)\,/\,\sqrt{s_1^2/n_1 + s_2^2/n_2}$ |
| Permutation p | proportion of $|\bar{d}^*| \geq |\bar{d}_{\text{obs}}|$ over 5000 shuffles |

With n = 7 and n = 6, normality is unverifiable. The permutation test only assumes exchangeability of labels under H0, a much weaker condition.

</div>


In [48]:
eng = np.array([95, 102, 88, 110, 98, 105, 92])  # Engineering salaries
mkt = np.array([78, 82, 75, 91, 85, 79])            # Marketing salaries

n_eng = len(eng)   # Engineering group size
n_mkt = len(mkt)   # Marketing group size

x_eng = np.mean(eng)  # Engineering sample mean
x_mkt = np.mean(mkt)  # Marketing sample mean

x_eng, x_mkt

(np.float64(98.57142857142857), np.float64(81.66666666666667))

In [49]:
obs_diff = x_eng - x_mkt  # observed difference in group means
obs_diff

np.float64(16.904761904761898)

In [50]:
var_eng = np.var(eng, ddof=1)  # Engineering sample variance
var_mkt = np.var(mkt, ddof=1)  # Marketing sample variance

se = np.sqrt(var_eng / n_eng + var_mkt / n_mkt)  # Welch standard error
se

np.float64(3.7173369640521416)

In [51]:
t_stat = obs_diff / se  # Welch t-statistic
t_stat

np.float64(4.547546286020462)

#### Part (a) : Welch's t-Test


In [52]:
alpha = 0.05

# Welch-Satterthwaite degrees of freedom
num = (var_eng / n_eng + var_mkt / n_mkt) ** 2
den = (var_eng / n_eng) ** 2 / (n_eng - 1) + (var_mkt / n_mkt) ** 2 / (n_mkt - 1)
df_welch = num / den

t_crit = t.ppf(1 - alpha / 2, df_welch)  # two-tailed critical value
t_crit

np.float64(2.2049627828302314)

**|t| > t_crit, Reject H0.**


In [53]:
p_val = 2 * t.sf(abs(t_stat), df_welch)  # two-tailed Welch p-value
p_val

np.float64(0.0008651210641434832)

In [54]:
margin = t_crit * se    # margin of error
lo = obs_diff - margin  # lower bound of 95% CI
hi = obs_diff + margin  # upper bound of 95% CI
lo, hi

(np.float64(8.708172247787804), np.float64(25.101351561735992))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (a) Conclusion**

t_stat (~4.55) > t_crit &nbsp;&nbsp;|&nbsp;&nbsp; p-value < 0.05

**Reject H0.** Engineering salaries are significantly higher than Marketing at α = 0.05.

</div>


#### Part (b) : Permutation Test


In [55]:
np.random.seed(42)  # fix seed for reproducibility

combined = np.concatenate([eng, mkt])  # pool all salaries together
n_total  = len(combined)               # total observations
n_perm   = 5000                        # number of permutations

perm_diffs = np.zeros(n_perm)  # store permuted mean differences

for i in range(n_perm):
    shuffled  = np.random.permutation(combined)    # randomly reassign all labels
    perm_eng  = shuffled[:n_eng]                   # first n_eng go to "Engineering"
    perm_mkt  = shuffled[n_eng:]                   # rest go to "Marketing"
    perm_diffs[i] = np.mean(perm_eng) - np.mean(perm_mkt)  # permuted difference

perm_diffs.shape

(5000,)

In [56]:
# two-tailed: count how many permuted diffs are at least as extreme as observed
p_val_perm = np.mean(np.abs(perm_diffs) >= np.abs(obs_diff))
p_val_perm

np.float64(0.0014)

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (b) Conclusion**

Permutation p-value < 0.05, Reject H0. Very few of 5,000 label shuffles produce a mean difference as large as 16.9.

</div>


#### Part (c) : Comparison


In [57]:
p_val, p_val_perm, obs_diff, df_welch  # t-test p, permutation p, observed diff, Welch df

(np.float64(0.0008651210641434832),
 np.float64(0.0014),
 np.float64(16.904761904761898),
 np.float64(10.839701568157425))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (c) Conclusion**

Both tests agree (reject H0). With n = 7 and n = 6 the permutation test is more trustworthy, it only requires exchangeability under H0, not normality.

</div>


#### Part (d) : What "Assumption-Free" Actually Means

Calling the permutation test "assumption-free" is a slight exaggeration. What it really means:

**What the permutation test does NOT assume:**
- No normality required within groups
- No specific distributional family (t, normal, exponential, etc.)
- No assumption about equal or unequal variances

**What the permutation test DOES assume:**
- **Exchangeability under H₀:** If salary is truly unrelated to department, then any permutation of labels is just as likely as any other. This is the core assumption.
- **Independence of observations:** Salaries are from different people and assumed independent.
- **Same distribution under H₀:** Both groups are drawn from the same distribution when H₀ is true (otherwise the labels wouldn't be truly exchangeable).

| Assumption | t-test | Permutation test |
|---|---|---|
| Normality within groups | Yes | No |
| Specific distribution family | Yes (t) | No |
| Equal/unequal variances (choose one) | Yes | No |
| Independence of observations | Yes | Yes |
| Exchangeability of labels under H₀ | Implicit | Yes (explicit) |

So the permutation test trades strong parametric assumptions (normality, t-distribution) for one weaker assumption (exchangeability). That's a good trade, especially at small sample sizes. The tradeoff is computational: 5,000 permutations instead of a formula. With modern hardware that's not a real concern.

**Part (d) Conclusion**

"Assumption-free" really means: free from distributional assumptions, not free from all assumptions. The permutation test's core assumption (exchangeability under H₀) is almost always reasonable when testing whether group membership matters. For small samples like n = 7 and n = 6, it's the more honest and robust choice.


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q8. Bayesian t-Test**

n = 12 observations, x̄ = 8.5, s = 2.1. H0: μ = 7.  
(a) Run the frequentist one-sample t-test.  
(b) Using the JZS Cauchy prior (scale = 0.707), compute BF10.  
(c) Interpret BF = 5, BF = 0.2, and BF = 0.03.  
(d) Name one thing a Bayes Factor can do that a p-value cannot.

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ = 7 &nbsp;|&nbsp; H1: μ ≠ 7, two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| t-statistic | $t = (\bar{x} - \mu_0)\,/\,(s/\sqrt{n})$, df = n-1 |
| BF10 | $P(\text{data}\mid H_1)\,/\,P(\text{data}\mid H_0)$ via numerical integration of JZS Cauchy prior |

BF10 > 1 favours H1; BF10 < 1 favours H0. Unlike p-values, BF can provide positive evidence for the null.

</div>


In [58]:
n  = 12    # sample size
x  = 8.5   # sample mean
s  = 2.1   # sample standard deviation
mu = 7     # null hypothesis mean
alpha = 0.05

#### Part (a) : Frequentist One-Sample t-Test


In [59]:
df = n - 1                  # degrees of freedom
se = s / n**0.5             # standard error of the mean
t_stat = (x - mu) / se     # one-sample t-statistic
t_stat

2.4743582965269675

In [60]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=11
t_crit

np.float64(2.200985160091639)

**|t| = 2.474 > t_crit = 2.201, Reject H0.**


In [61]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.030881561145156822)

**p-value = 0.0311 < 0.05, Reject H0.**


In [62]:
se_val = s / np.sqrt(n)       # standard error
lo = x - t_crit * se_val      # lower bound of 95% CI
hi = x + t_crit * se_val      # upper bound of 95% CI
lo, hi

(np.float64(7.165723656605657), np.float64(9.834276343394343))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (a) Conclusion**

t_stat (2.474) > t_crit (2.201) &nbsp;&nbsp;|&nbsp;&nbsp; p-value (0.0311) < alpha (0.05)

**Reject H0.** Sample mean of 8.5 is significantly different from null value of 7; 95% CI (7.16, 9.84) lies entirely above 7.

</div>


#### Part (b) : Bayes Factor BF₁₀ via JZS Cauchy Prior

**Why the JZS prior**

The JZS (Jeffreys-Zellner-Siow) prior places a Cauchy distribution on the standardised effect size δ = (μ − μ₀)/σ, with scale r = 0.707 (= 1/√2). This is a default prior: it's symmetric around zero (no a priori direction assumed), has heavy tails (allows large effects), and is invariant under rescaling. The Bayes Factor BF₁₀ is the ratio of marginal likelihoods P(data | H₁) / P(data | H₀). Computing it requires integrating the Cauchy prior over all possible effect sizes.

**Formula**

$$BF_{10} = \frac{\int_0^\infty \left(1 + \frac{n\delta^2}{t^2 + n - 1}\right)^{-n/2} \cdot p(\delta \mid r) \, d\delta}{(1 + t^2/(n-1))^{-n/2}}$$

In practice this integral is computed numerically. The `pingouin` library provides a direct implementation.


In [63]:
r = 1 / np.sqrt(2)  # JZS Cauchy prior scale

BF10, _ = quad(
    lambda g: (
        g**(-3/2) * np.exp(-1/(2*g)) / np.sqrt(2*np.pi) *
        (1 + n*g*r**2)**(-1/2) *
        (1 + t_stat**2 / (df*(1 + n*g*r**2)))**(-(df+1)/2) /
        (1 + t_stat**2 / df)**(-(df+1)/2)
    ),
    0, np.inf
)

BF10  # Bayes Factor: ratio of marginal likelihoods P(data|H1) / P(data|H0)

2.4068711107202443

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (b) Conclusion**

BF10 ≈ 2.41. Data are ~2.4x more likely under H1 than H0, anecdotal evidence for H1 (Jeffreys scale). More cautious than the frequentist rejection at α = 0.05.

</div>


#### Part (c) : Interpreting Specific Bayes Factors

**Jeffreys' Interpretation Scale**

| BF₁₀ | Interpretation |
|---|---|
| > 100 | Extreme evidence for H₁ |
| 30–100 | Very strong evidence for H₁ |
| 10–30 | Strong evidence for H₁ |
| 3–10 | Moderate evidence for H₁ |
| 1–3 | Anecdotal evidence for H₁ |
| 1 | No evidence either way |
| 1/3–1 | Anecdotal evidence for H₀ |
| 1/10–1/3 | Moderate evidence for H₀ |
| < 1/10 | Strong evidence for H₀ |

**BF₁₀ = 5** → The data are 5 times more likely under H₁ than H₀. This is moderate evidence in favour of H₁. Worth noting, but a replication would strengthen the conclusion.

**BF₁₀ = 0.2 = 1/5** → The data are 5 times more likely under H₀ than H₁. This is moderate evidence in favour of H₀, the data are actively supporting the null, not just failing to reject it.

**BF₁₀ = 0.03 ≈ 1/33** → The data are approximately 33 times more likely under H₀ than H₁. This is strong-to-very-strong evidence for H₀. The null hypothesis is strongly supported.

**Part (c) Conclusion**

| BF₁₀ | Numeric meaning | Verbal interpretation |
|---|---|---|
| 5 | H₁ is 5× more likely | Moderate evidence for H₁ |
| 0.2 | H₀ is 5× more likely | Moderate evidence for H₀ |
| 0.03 | H₀ is ~33× more likely | Strong evidence for H₀ |

Unlike the p-value, Bayes Factors are symmetric: they can support either hypothesis, and their magnitude quantifies the strength of that support on a continuous scale.


#### Part (d) : What a Bayes Factor Can Do That a p-value Cannot

**A Bayes Factor can quantify evidence in favour of the null hypothesis. A p-value fundamentally cannot.**

A p-value answers: *"How surprising is this data if H₀ is true?"* A large p-value (e.g. p = 0.8) does not mean H₀ is supported, it just means the data are not inconsistent with H₀. Failing to reject is not evidence for the null; it could simply reflect low power, a small sample, or high variance.

A Bayes Factor answers: *"How much more likely is the data under H₁ versus H₀?"* When BF₁₀ = 0.03, the data are actively supporting H₀, 33 times more likely under H₀ than H₁. That's a positive, quantifiable statement about the null, not a silence.

**Part (d) Conclusion**

| | p-value | Bayes Factor |
|---|---|---|
| Can reject H₁ | Yes (implicitly, via rejection) | Yes |
| Can support H₀ | **No** | **Yes** |
| Quantifies evidence strength | No | Yes |
| Depends on stopping rule | Yes | No |
| Requires prior | No | Yes |

This is the single most important practical difference. In fields where demonstrating that an effect is absent matters (e.g. safety testing, bioequivalence, cognitive neuroscience null results), the Bayes Factor is the appropriate tool. The p-value cannot distinguish "no evidence for an effect" from "evidence for no effect", the Bayes Factor can.


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q9. Hotelling's T² (Multivariate Extension)**

20 patients measured on blood pressure and cholesterol before/after treatment.  
Mean differences: d̄_BP = 8.2, d̄_Chol = 12.1 &nbsp;|&nbsp; S = [[36, 18], [18, 64]]  
(a) Compute T² = n · d̄ᵀ · S⁻¹ · d̄  
(b) Convert to F: F = T² · (n-p) / [p · (n-1)]  
(c) Why is this better than running two separate paired t-tests?

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ_BP = 0 and μ_Chol = 0 simultaneously  
H1: at least one mean difference is non-zero, omnibus test at α = 0.05

| Step | Formula |
|------|---------|
| T² | $T^2 = n\,\bar{\mathbf{d}}^\top S^{-1} \bar{\mathbf{d}}$ |
| F conversion | $F = T^2(n-p)\,/\,[p(n-1)]$, df1 = p, df2 = n-p |

BP and cholesterol are correlated outcomes. Two separate t-tests inflate Type I error and ignore the covariance; T² tests both jointly while controlling α at 0.05.

</div>


In [64]:
n     = 20                                    # number of patients
p     = 2                                     # number of outcome variables
d_bar = np.array([8.2, 12.1])                # vector of mean differences [BP, Chol]
S     = np.array([[36, 18], [18, 64]])        # covariance matrix of differences
S_inv = np.linalg.inv(S)                     # inverse of covariance matrix
alpha = 0.05                                  # significance level

T2 = n * d_bar @ S_inv @ d_bar               # Hotelling's T² statistic
T2

np.float64(60.62828282828283)

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (a) Conclusion**

T² = 60.63. Summarises how far the mean difference vector is from zero, weighted by the inverse covariance matrix.

</div>


#### Part (b) : Converting T² to F


In [65]:
f_stat = T2 * (n - p) / (p * (n - 1))     # convert T² to F-statistic
f_stat

np.float64(28.718660287081338)

In [66]:
df1    = p                                  # numerator df
df2    = n - p                              # denominator df
f_crit = f.ppf(1 - alpha, df1, df2)        # one-tailed upper critical value (F is always right-tailed)
f_crit

np.float64(3.5545571456617866)

**F = 28.719 > F_crit = 3.555, Reject H0.**


In [67]:
p_val = f.sf(f_stat, df1, df2)  # p-value from F-distribution
p_val

np.float64(2.5071265475037957e-06)

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (b) Conclusion**

F (28.72) >> F_crit (3.56) &nbsp;&nbsp;|&nbsp;&nbsp; p-value = 0.0000025 < alpha (0.05)

**Reject H0.** Very strong evidence that treatment produced a significant simultaneous change in at least one outcome.

</div>


#### Part (c) : Why Hotelling's T² Is Better Than Two Separate t-Tests


In [68]:
se_BP   = np.sqrt(36) / np.sqrt(n)            # SE for blood pressure
se_Chol = np.sqrt(64) / np.sqrt(n)            # SE for cholesterol

t_BP   = 8.2  / se_BP                         # t-statistic for BP
t_Chol = 12.1 / se_Chol                       # t-statistic for cholesterol

p_BP   = 2 * t.sf(abs(t_BP),   n - 1)         # two-tailed p-value for BP
p_Chol = 2 * t.sf(abs(t_Chol), n - 1)         # two-tailed p-value for cholesterol

corr_BP_Chol = 18 / (np.sqrt(36) * np.sqrt(64))  # correlation between the two outcomes

p_BP, p_Chol, corr_BP_Chol

(np.float64(7.086054685196583e-06),
 np.float64(1.8416973242238146e-06),
 np.float64(0.375))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (c) Conclusion**

| | Two Separate t-Tests | Hotelling's T² |
|---|---|---|
| Number of tests | 2 | 1 |
| Type I error control | Inflated | Exactly α = 0.05 |
| Uses outcome correlation | No | Yes (via S⁻¹) |

Separate t-tests inflate Type I error and ignore the covariance of 18 (r ≈ 0.375). T² provides one coherent decision that correctly controls the family-wise error rate.

</div>


<div style="background-color:#dbeafe; padding:15px; border-radius:8px; border-left:5px solid #2563eb;">

### **Q10. Paired t-Test on Real Data with Sensitivity Analysis**

Before: 62, 55, 68, 71, 59, 63, 66, 58, 61, 64, 57, 60  
After: 70, 60, 75, 78, 65, 72, 74, 64, 69, 73, 63, 68  
(a) Perform a paired t-test at α = 0.05.  
(b) Construct a 95% CI for the mean improvement.  
(c) One student's after-score changes from 75 to 95. How does this affect t and p?  
(d) When should a Wilcoxon signed-rank test be used instead?

</div>


<div style="background-color:#f8fafc; padding:13px; border-radius:6px; border-left:4px solid #2563eb; margin-bottom:8px;">

**Hypotheses:**  
H0: μ_d = 0 (training has no effect)  
H1: μ_d ≠ 0, two-tailed at α = 0.05

| Step | Formula |
|------|---------|
| Differences | $d_i = \text{after}_i - \text{before}_i$ |
| t-statistic | $t = \bar{d}\,/\,(s_d/\sqrt{n})$, df = n-1 |
| 95% CI | $\bar{d} \pm t_{\alpha/2,\,df} \cdot SE_d$ |

Same 12 students measured twice, paired design removes between-student variability and is more powerful than independent-samples for this structure.

</div>


In [69]:
before = np.array([62, 55, 68, 71, 59, 63, 66, 58, 61, 64, 57, 60])  # memory scores before training
after  = np.array([70, 60, 75, 78, 65, 72, 74, 64, 69, 73, 63, 68])  # memory scores after training

n     = len(before)    # number of pairs
diff  = after - before  # individual differences
alpha = 0.05
df    = n - 1

diff

array([8, 5, 7, 7, 6, 9, 8, 6, 8, 9, 6, 8])

#### Part (a) : Paired t-Test


In [70]:
d_bar = np.mean(diff)          # mean of differences
s_d   = np.std(diff, ddof=1)   # sample SD of differences
se    = s_d / np.sqrt(n)       # standard error of the mean difference

t_stat = d_bar / se  # paired t-statistic
t_stat

np.float64(19.49815586643467)

In [71]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=11
t_crit

np.float64(2.200985160091639)

**|t| = 19.498 > t_crit = 2.201, Reject H0.**


In [72]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(7.018081387964999e-10)

**p-value ≈ 7.02 × 10⁻¹⁰ < 0.05, Reject H0.**


In [73]:
t_scipy, p_scipy = ttest_rel(after, before)  # scipy paired t-test verification
t_scipy, p_scipy

(np.float64(19.498155866434672), np.float64(7.018081387964983e-10))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (a) Conclusion**

t_stat (19.498) >> t_crit (2.201) &nbsp;&nbsp;|&nbsp;&nbsp; p-value ≈ 7e-10 < alpha (0.05)

**Reject H0.** Every student improved; mean gain = 7.25 points. Highly significant.

</div>


#### Part (b) : 95% Confidence Interval


In [74]:
lo = d_bar - t_crit * se  # lower bound of 95% CI
hi = d_bar + t_crit * se  # upper bound of 95% CI
lo, hi

(np.float64(6.431607608433678), np.float64(8.068392391566322))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (b) Conclusion**

95% CI for mean improvement: **(6.43, 8.07)**, entirely above zero. Confident the program improved scores by 6.43 to 8.07 points.

</div>


#### Part (c) : Sensitivity Analysis, Changing One Score from 75 to 95

Student 3's after-score changes from 75 to 95, making their difference jump from 7 to 27. This is a single influential observation. The mean difference will increase, but more importantly, the variance of the differences will inflate dramatically because 27 is now a large outlier relative to the rest of the differences (all between 5 and 9).


In [75]:
after_mod    = after.copy()
after_mod[2] = 95                          # change student 3's after-score

diff_mod  = after_mod - before             # new differences
d_bar_mod = np.mean(diff_mod)              # new mean difference
s_d_mod   = np.std(diff_mod, ddof=1)       # new SD (much larger due to outlier)
se_mod    = s_d_mod / np.sqrt(n)           # new standard error
t_mod     = d_bar_mod / se_mod             # new t-statistic
p_mod     = 2 * t.sf(abs(t_mod), df)      # new p-value

d_bar_mod, s_d_mod, t_mod, p_mod

(np.float64(8.916666666666666),
 np.float64(5.838093296045665),
 np.float64(5.29080948784539),
 np.float64(0.0002559969744402437))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (c) Conclusion**

| | Original | Modified (75 -> 95) |
|---|---|---|
| Mean difference | 7.25 | 8.92 |
| SD of differences | 1.29 | 5.84 |
| t-statistic | 19.50 | 5.29 |
| p-value | ~7e-10 | 0.0003 |

**Still reject H0**, but the single outlier inflates SD 4.5x, shrinking t from 19.50 to 5.29. Demonstrates why outliers in differences matter in small paired samples.

</div>


#### Part (d) : When to Use the Wilcoxon Signed-Rank Test Instead

The paired t-test assumes that the differences are approximately normally distributed. When that assumption is violated, the Wilcoxon signed-rank test is the appropriate non-parametric alternative. It ranks the absolute differences and uses the sign information without relying on normality.

Use the Wilcoxon test when:

- The differences are clearly non-normal, especially with heavy tails or strong skew
- The sample is very small (n < 10) and normality cannot be assessed
- There are outliers in the differences that would distort the mean (as seen in part (c))
- The data are ordinal rather than continuous (e.g. ranked scores, Likert scales)
- A single influential observation is suspected of driving the t-test result

The tradeoff is power: when the data are genuinely normal, the Wilcoxon test is slightly less powerful than the paired t-test (about 95% relative efficiency). But when normality fails, the t-test's Type I error and power guarantees break down, and the Wilcoxon test is more reliable.


#### Wilcoxon Signed-Rank Test (Original Data)


In [76]:
w_stat, p_wilcox = wilcoxon(after, before, alternative='two-sided')  # Wilcoxon signed-rank test
w_stat, p_wilcox, p_val  # W statistic, Wilcoxon p, paired t p

(np.float64(0.0), np.float64(0.00048828125), np.float64(7.018081387964999e-10))

<div style="background-color:#f0fdf4; padding:12px; border-radius:6px; border-left:4px solid #16a34a;">

**Part (d) Conclusion**

Wilcoxon agrees with the paired t-test here (both reject H0). Use Wilcoxon when: differences are non-normal, n < 10, outliers are present in differences, or data are ordinal.

</div>
